aws/external/resolve_credentials.go

```diff
@@ -0,0 +1,404 @@
+package external
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	"github.com/aws/aws-sdk-go-v2/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go-v2/aws/ec2rolecreds"
+	"github.com/aws/aws-sdk-go-v2/aws/endpointcreds"
+	"github.com/aws/aws-sdk-go-v2/aws/processcreds"
+	"github.com/aws/aws-sdk-go-v2/aws/stscreds"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+const (
+	// valid credential source values
+	credSourceEc2Metadata  = "Ec2InstanceMetadata"
+	credSourceEnvironment  = "Environment"
+	credSourceECSContainer = "EcsContainer"
+)
+
+var (
+	ecsContainerEndpoint = "http://169.254.170.2" // not constant to allow for swapping during unit-testing
+
+	// WebIdentityEmptyRoleARNErr will occur if 'AWS_WEB_IDENTITY_TOKEN_FILE' was set but
+	// 'AWS_ROLE_ARN' was not set.
+	WebIdentityEmptyRoleARNErr = awserr.New(stscreds.ErrCodeWebIdentity, "role ARN is not set", nil)
+
+	// WebIdentityEmptyTokenFilePathErr will occur if 'AWS_ROLE_ARN' was set but
+	// 'AWS_WEB_IDENTITY_TOKEN_FILE' was not set.
+	WebIdentityEmptyTokenFilePathErr = awserr.New(stscreds.ErrCodeWebIdentity, "token file path is not set", nil)
```

Comment:
Probably don't want to expose error values because we'll never be able to change them to errors that can be includes metadata about the filename that failed, or other attributes about the error that may be useful to return.